Compiler crashes when value class extends an alias of `AnyVal`

## Compiler version

https://github.com/scala/scala3/commit/cc38962e04a842695647f7006384e18e59ec6c56

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```scala
type AliasToAnyVal = AnyVal
class Foo(a: Int) extends AliasToAnyVal
```

## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
scalac anyval.scala

  unhandled exception while running erasure on anyval.scala

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/scala/scala3/issues/new/choose
  For non-enriched exceptions, compile with -Xno-enrich-error-messages.

     while compiling: anyval.scala
        during phase: erasure
                mode: Mode(ImplicitsEnabled)
     library version: version 2.13.14
    compiler version: version 3.5.0
            settings:

Exception in thread "main" java.lang.AssertionError: NoDenotation.owner
	at dotty.tools.dotc.core.SymDenotations$NoDenotation$.owner(SymDenotations.scala:2623)
	at dotty.tools.dotc.ast.tpd$.ref(tpd.scala:462)
	at dotty.tools.dotc.transform.Erasure$Boxing$.wrap$1(Erasure.scala:338)
	at dotty.tools.dotc.transform.Erasure$Boxing$.cast(Erasure.scala:366)
	at dotty.tools.dotc.transform.Erasure$Boxing$.unbox(Erasure.scala:320)
	at dotty.tools.dotc.transform.Erasure$Boxing$.adaptToType(Erasure.scala:403)
	at dotty.tools.dotc.transform.Erasure$Typer.adapt(Erasure.scala:1098)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3554)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3558)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3669)
	at dotty.tools.dotc.transform.Erasure$Typer.$anonfun$7(Erasure.scala:863)
	at dotty.tools.dotc.core.Decorators$.zipWithConserve(Decorators.scala:160)
	at dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:863)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3392)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3477)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3554)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3558)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3669)
	at dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:850)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3392)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3477)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3554)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3558)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3669)
	at dotty.tools.dotc.typer.Typer.$anonfun$64(Typer.scala:2816)
	at dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
	at dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2816)
	at dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:967)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3374)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3476)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3554)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3558)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3580)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3626)
	at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1080)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:3074)
	at dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1056)
	at dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:3380)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3384)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3476)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3554)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3558)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3580)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3626)
	at dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1080)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:3207)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3426)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3477)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3554)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:3558)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3669)
	at dotty.tools.dotc.transform.Erasure.run(Erasure.scala:145)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:380)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:373)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:343)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:336)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:384)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:396)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:396)
	at dotty.tools.dotc.Run.compileSources(Run.scala:282)
	at dotty.tools.dotc.Run.compile(Run.scala:267)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
	at dotty.tools.dotc.Driver.process(Driver.scala:201)
	at dotty.tools.dotc.Driver.process(Driver.scala:169)
	at dotty.tools.dotc.Driver.process(Driver.scala:181)
	at dotty.tools.dotc.Driver.main(Driver.scala:211)
	at dotty.tools.MainGenericCompiler$.run$1(MainGenericCompiler.scala:162)
	at dotty.tools.MainGenericCompiler$.main(MainGenericCompiler.scala:186)
	at dotty.tools.MainGenericCompiler.main(MainGenericCompiler.scala)
```
</details>

## Notes

The check to determine if a class is a value class doesn't take aliasing into account. See:

https://github.com/scala/scala3/blob/cc38962e04a842695647f7006384e18e59ec6c56/compiler/src/dotty/tools/dotc/ast/Desugar.scala#L597-L601

--- a/compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/ErrorMessageID.scala
@@ -223,6 +223,7 @@ enum ErrorMessageID(val isActive: Boolean = true) extends java.lang.Enum[ErrorMe
   case IllegalUnrollPlacementID // errorNumber: 207
   case ExtensionHasDefaultID // errorNumber: 208
   case FormatInterpolationErrorID // errorNumber: 209
+  case ValueClassCannotExtendAliasOfAnyValID // errorNumber: 210
 
   def errorNumber = ordinal - 1
 

--- a/compiler/src/dotty/tools/dotc/reporting/messages.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/messages.scala
@@ -1813,6 +1813,12 @@ class ValueClassParameterMayNotBeCallByName(valueClass: Symbol, param: Symbol)(u
   def explain(using Context) = ""
 }
 
+class ValueClassCannotExtendAliasOfAnyVal(valueClass: Symbol, alias: Symbol)(using Context)
+  extends SyntaxMsg(ValueClassCannotExtendAliasOfAnyValID) {
+  def msg(using Context) = i"""A value class cannot extend a type alias ($alias) of ${hl("AnyVal")}"""
+  def explain(using Context) = ""
+}
+
 class SuperCallsNotAllowedInlineable(symbol: Symbol)(using Context)
   extends SyntaxMsg(SuperCallsNotAllowedInlineableID) {
   def msg(using Context) = i"Super call not allowed in inlineable $symbol"

--- a/compiler/src/dotty/tools/dotc/transform/TreeChecker.scala
+++ b/compiler/src/dotty/tools/dotc/transform/TreeChecker.scala
@@ -244,7 +244,7 @@ object TreeChecker {
     private val everDefinedSyms = MutableSymbolMap[untpd.Tree]()
 
     // don't check value classes after typer, as the constraint about constructors doesn't hold after transform
-    override def checkDerivedValueClass(clazz: Symbol, stats: List[Tree])(using Context): Unit = ()
+    override def checkDerivedValueClass(cdef: untpd.TypeDef, clazz: Symbol, stats: List[Tree])(using Context): Unit = ()
 
     def withDefinedSyms[T](trees: List[untpd.Tree])(op: => T)(using Context): T = {
       var locally = List.empty[Symbol]

--- a/compiler/src/dotty/tools/dotc/typer/Checking.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Checking.scala
@@ -743,7 +743,7 @@ object Checking {
   }
 
   /** Verify classes extending AnyVal meet the requirements */
-  def checkDerivedValueClass(clazz: Symbol, stats: List[Tree])(using Context): Unit = {
+  def checkDerivedValueClass(cdef: untpd.TypeDef, clazz: Symbol, stats: List[Tree])(using Context): Unit = {
     def checkValueClassMember(stat: Tree) = stat match {
       case _: TypeDef if stat.symbol.isClass =>
         report.error(ValueClassesMayNotDefineInner(clazz, stat.symbol), stat.srcPos)
@@ -760,6 +760,15 @@ object Checking {
     // enum extending a value class type (AnyVal or an alias of it)
     // The error message 'EnumMayNotBeValueClassesID' will take care of generating the error message (See #22236)
     if (clazz.isDerivedValueClass && !clazz.isEnumAnonymClass) {
+      val parentOpt = cdef.rhs match {
+        case impl: Template =>
+          impl.parents.headOption
+        case _ => None
+      }
+      val isExtendingAliasOfAnyVal = parentOpt.exists { parent =>
+        parent.symbol.isAliasType && parent.tpe.nn.dealias =:= defn.AnyValType
+      }
+
       if (clazz.is(Trait))
         report.error(CannotExtendAnyVal(clazz), clazz.srcPos)
       if clazz.is(Module) then
@@ -770,6 +779,8 @@ object Checking {
         report.error(ValueClassesMayNotBeAbstract(clazz), clazz.srcPos)
       if (!clazz.isStatic)
         report.error(ValueClassesMayNotBeContainted(clazz), clazz.srcPos)
+      if (isExtendingAliasOfAnyVal)
+        report.error(ValueClassCannotExtendAliasOfAnyVal(clazz, parentOpt.get.symbol), clazz.srcPos)
       if (isDerivedValueClass(underlyingOfValueClass(clazz.asClass).classSymbol))
         report.error(ValueClassesMayNotWrapAnotherValueClass(clazz), clazz.srcPos)
       else {
@@ -1307,8 +1318,8 @@ trait Checking {
     else tpt
 
   /** Verify classes extending AnyVal meet the requirements */
-  def checkDerivedValueClass(clazz: Symbol, stats: List[Tree])(using Context): Unit =
-    Checking.checkDerivedValueClass(clazz, stats)
+  def checkDerivedValueClass(cdef: untpd.TypeDef, clazz: Symbol, stats: List[Tree])(using Context): Unit =
+    Checking.checkDerivedValueClass(cdef, clazz, stats)
 
   /** Check that case classes are not inherited by case classes.
    */
@@ -1689,7 +1700,7 @@ trait NoChecking extends ReChecking {
   override def checkNoTargetNameConflict(stats: List[Tree])(using Context): Unit = ()
   override def checkParentCall(call: Tree, caller: ClassSymbol)(using Context): Unit = ()
   override def checkSimpleKinded(tpt: Tree)(using Context): Tree = tpt
-  override def checkDerivedValueClass(clazz: Symbol, stats: List[Tree])(using Context): Unit = ()
+  override def checkDerivedValueClass(cdef: untpd.TypeDef, clazz: Symbol, stats: List[Tree])(using Context): Unit = ()
   override def checkCaseInheritance(parentSym: Symbol, caseCls: ClassSymbol, pos: SrcPos)(using Context): Unit = ()
   override def checkNoForwardDependencies(vparams: List[ValDef])(using Context): Unit = ()
   override def checkMembersOK(tp: Type, pos: SrcPos)(using Context): Type = tp

--- a/compiler/src/dotty/tools/dotc/typer/Checking.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Checking.scala
@@ -756,19 +756,18 @@ object Checking {
       case _ =>
         report.error(ValueClassesMayNotContainInitalization(clazz), stat.srcPos)
     }
+    inline def checkParentIsNotAnyValAlias(): Unit =
+      cdef.rhs match {
+        case impl: Template =>
+          val parent = impl.parents.head
+          if parent.symbol.isAliasType && parent.typeOpt.dealias =:= defn.AnyValType then
+            report.error(ValueClassCannotExtendAliasOfAnyVal(clazz, parent.symbol), cdef.srcPos)
+        case _ => ()
+      }
     // We don't check synthesised enum anonymous classes that are generated from
     // enum extending a value class type (AnyVal or an alias of it)
     // The error message 'EnumMayNotBeValueClassesID' will take care of generating the error message (See #22236)
     if (clazz.isDerivedValueClass && !clazz.isEnumAnonymClass) {
-      val parentOpt = cdef.rhs match {
-        case impl: Template =>
-          impl.parents.headOption
-        case _ => None
-      }
-      val isExtendingAliasOfAnyVal = parentOpt.exists { parent =>
-        parent.symbol.isAliasType && parent.tpe.nn.dealias =:= defn.AnyValType
-      }
-
       if (clazz.is(Trait))
         report.error(CannotExtendAnyVal(clazz), clazz.srcPos)
       if clazz.is(Module) then
@@ -779,8 +778,9 @@ object Checking {
         report.error(ValueClassesMayNotBeAbstract(clazz), clazz.srcPos)
       if (!clazz.isStatic)
         report.error(ValueClassesMayNotBeContainted(clazz), clazz.srcPos)
-      if (isExtendingAliasOfAnyVal)
-        report.error(ValueClassCannotExtendAliasOfAnyVal(clazz, parentOpt.get.symbol), clazz.srcPos)
+
+      checkParentIsNotAnyValAlias()
+
       if (isDerivedValueClass(underlyingOfValueClass(clazz.asClass).classSymbol))
         report.error(ValueClassesMayNotWrapAnotherValueClass(clazz), clazz.srcPos)
       else {

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -3243,7 +3243,7 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
       checkNonCyclicInherited(cls.thisType, cls.info.parents, cls.info.decls, cdef.srcPos)
 
       // check value class constraints
-      checkDerivedValueClass(cls, body1)
+      checkDerivedValueClass(cdef, cls, body1)
 
       val effectiveOwner = cls.owner.skipWeakOwner
       if cls.is(ModuleClass)

--- a/tests/neg/i21918.scala
+++ b/tests/neg/i21918.scala
@@ -1,0 +1,2 @@
+type AliasToAnyVal = AnyVal
+class Foo(a: Int) extends AliasToAnyVal // error

